nimbus/src/lib.rs

```diff
@@ -80,7 +82,23 @@ impl NimbusClient {
             .map(|e| e.branch_slug.clone()))
     }
 
+    pub fn get_global_user_participation(&self) -> Result<Option<bool>> {
+        self.db()?.get(StoreId::Meta, DB_KEY_GLOBAL_USER_PARTICIPATION)
+    }
+
+    pub fn set_global_user_participation(&self, flag: bool) -> Result<()> {
+        self.db()?.put(StoreId::Meta, DB_KEY_GLOBAL_USER_PARTICIPATION, &flag)
```

Comment:
I imagine that setting this to false should "do something" too? I raised this in slack but it got lost in other noise, but we should possibly clarify the requirements here. The general question is "does opting out then opting back in do a full reset?" Specifically:

* On a global opt-out then opt back in, do I reuse my nimbus-id? If so, that implies I'll probably be re-enrolled in the exact same experiments.
* If I opt out of an experiment, then opt out globally, then opt in globally, am I again eligible for that first experiment?

Reply:
For some context, Glean rotates it's client id every time the user opts out, so I am leaning towards rotating the `nimbus_id` for the reasons @mhammond mentions. Unfortunately, that might make it difficult to track that a user was enrolled but "opted out" of an experiment via telemetry, which sounds like it will be a requirement for analysis purposes.

Reply:
These are good points, both.

I'm not clear on the loose ends need tying up when we're telling anyone we're no longer enrolled in.

My understanding is that the `nimbus_id` is only used internally to nimbus, and never leaves the device.

 * what happens when the user opts out?
     - do we need to tell anyone/Glean that this device is no longer enrolled in any experiments? AFAICT this will be picked up next ping that is sent, that this `client_id` is enrolled in no experiments. Is this soon enough, or should we do that immediately?
 * what happens to when the user re-opts in? 
    - Do we need to tell anyone/Glean that this devices is now enrolled in a set of experiments? This is complicated by the fact that Glean and Nimbus don't meet anywhere where Nimbus can call Glean directly.
     - I suspect that the re-bucketing would not be desirable. From a UX point-of-view, the user would be rolling the dice of which feature they get each time they ticked and unticked a box. I don't know if there are subtle statistics problems if you don't do that rebucketing, though.

@mhammond @travis79 what say you?

Reply:
I believe that the idea is to record these changes of state as events in Glean, so if the user opts out, or opts back in, we can record the exact moment and state.  In the case here, I don't think opting out and back in should land you in a different bucket or branch, in fact, I think that once a user has opted out, that they should not be allowed to opt back into the same experiment.  With the effect that this might have on analysis, it might be a good question to get data-science input on for what the desired behavior is.

Reply:
>  I don't think opting out and back in should land you in a different bucket or branch

That's a good point and would argue for keeping the same ID.

>  think that once a user has opted out, that they should not be allowed to opt back into the same experiment.

User's will not be able to opt-in - but as above, if we change their ID they might now qualify - IOW, changing their ID gives them a back-door into opting-in - they just keep opting out them back in again until their new ID qualifies them.

>  it might be a good question to get data-science input

Yes, exactly - I don't think our opinions are particularly relevant :)

Reply:
> what happens when the user opts out? 

Tim's document on [unpacking unenrollment](https://docs.google.com/document/d/18Vdn4zZX2D1u9AqzHRY0lUSUbpBdouqWtVp3Bu8vYQA/) seems relevant here, particularly:

"""
Any client that enrolls in the experiment and is assigned to a branch should have their data included for the assigned condition in the final analysis, even if they withdraw and subsequently have a different user experience than they were assigned
"""

By this logic, if a user opts out *after having been enrolled in an experiment*, then we should continue to report that experiment and branch in their glean metrics until such time as the experiment ends. We should not actually show them the experimental treatment, because they've opted out of the experiment. But the fact that they were in the experiment at some point is relevant to analysis.  (And if they don't want us to report on this at all, they can disable telemetry).

Implicit in this requirement is that we remember a user's past enrollments, rather than deriving them afresh each time using the nimbus_id. So I don't think changing nimbus_id should have any effect on for experiment in which they'd previously been enrolled (and also, it's not clear to me what the intended effect of changing it would be).

Reply:
> Implicit in this requirement is that we remember a user's past enrollments, rather than deriving them afresh each time using the nimbus_id. So I don't think changing nimbus_id should have any effect on for experiment in which they'd previously been enrolled (and also, it's not clear to me what the intended effect of changing it would be).

I think the `nimbus_id` is an implementation detail of how the enrollment occurs; as long as it's deterministic, I don't think we need to store the past enrollment.

If the experiment changes (e.g. branches are removed) while the user is opted out, and then the user opts back in, then we'll be in trouble either way: by remembering the enrollment the now defunct branch and reason, OR enrolling into a new branch, but I think this is for a different issue to solve.

Reply:
> but I think this is for a different issue to solve.

Agreed